### PR TITLE
Docker setup

### PR DESCRIPTION
**Pull request by machour**
This is a first, lousy, attempt at implementing a docker setup for mysql tests. (https://github.com/yiisoft/yii-core/issues/157)

I only used docker on two exotic apps I'm maintaining, so I'm really not that skilled at it and someone else would probably come up with something more efficient.

Composer update & tests are running slow, we probably need something like docker-sync to have native-like speed.

You still need to change the database hostname to `db` in `vendor/yiisoft/yii-core/config/tests/params.php`

### CODE REVIEW

**Comment by machour** [comment]
cc @schmunk42

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+version: "3"
+
+services:
+  db:
+    image: mysql
+    container_name: yiisoft_db_mysql_db
```
Comment (schmunk42) [inline docker/docker-compose.yml]: I'd remove `container_name` you usually don't need that.

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    command: --default-authentication-plugin=mysql_native_password
+    volumes:
+      - ./mysql/setup:/docker-entrypoint-initdb.d
+    restart: always
```
Comment (schmunk42) [inline docker/docker-compose.yml]: Also not needed for development

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    build: ./php
+    container_name: yiisoft_db_mysql_php
+    depends_on:
+      - db
+    volumes:
+      - ./../:/var/www/html/
+    ports:
+      - "8100:80"
+    stdin_open: true
+    tty: true
```
Comment (schmunk42) [inline docker/docker-compose.yml]: Is there a special reason for the above two params? I actually did know them :)

File: docker/php/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FROM php:7.3-cli
```
Comment (schmunk42) [inline docker/php/Dockerfile]: Could you use the images from https://github.com/yiisoft/yii-docker - I think there's no need for a `Dockerfile` atm.

Comment (machour) [reply docker/php/Dockerfile]: I thought about it, but these would set up so much stuff not needed for every repository (imagemagick, nodejs, ..)

File: docker/docker-compose.yml (hunk 3 above)
Comment (machour) [reply docker/docker-compose.yml]: copy/pasta

File: docker/php/Dockerfile (hunk 4 above)
Comment (schmunk42) [reply docker/php/Dockerfile]: These should be the requirements for core modules (`requirements.php`), were there changes in Yii3 for this?

I know that node is debatable but in the long run I think it'll become a requirement like composer (at least for web apps).

Having the stuff ready in an image is one of the main purposes of Docker, IMO.

Comment (machour) [reply docker/php/Dockerfile]: All righty. I was aiming at a minimalistic setup for each repo requirements.
I'll update the PR to use the yii2 image.

Comment (machour) [reply docker/php/Dockerfile]: Done

**Comment by machour** [comment]
Still need to address:
- [ ] Perfs issue
- [ ] Automatic configuration of test parameters

I need guidance for both.

**Comment by schmunk42** [comment]
> Perfs issue

You need this https://github.com/yiisoft/yii-dev/blob/master/docker-compose.yml#L9 - reuse composer caches via host-volume. Will be slow for the first run.

> Automatic configuration of test parameters

Shouldn't the config-plugin be capable of that? Eg. using ENV vars. CC: @hiqsol 

> The docker folder name and location (I don't like that it's in the repo root, might confuse users)

Put it into `tests`, would make it clear what it is intended for.

**Comment by machour** [comment]
Did everything expect the config part.

When using an image in docker-compose, can you also make it `RUN` a command after the build?

If so, simply copying this `config.local.php` file to /vendor/yiisoft/yii-core/config/tests/ should do the trick:

```php
<?php

$config['database']['mysql]['dsn'] = 'mysql:host=db;dbname=yiitest';
```

**Comment by samdark** [comment]
Configs look good. How about a readme?

File: tests/docker/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+This docker setup allows you to run tests for this repository in a pre-configured environement.
+
+Start the containers and get a shell into the PHP container:
+```shell
+$ docker-compose up --build -d
+Recreating docker_db_1 ... done
+Recreating docker_web_1 ... done
+
+$ docker exec -it docker_web_1 bash
```
Comment (schmunk42) [inline tests/docker/README.md]: Would change this to `docker-compose exec web bash`

Comment (machour) [reply tests/docker/README.md]: Yay, learned something!

**Pull request merged**